fivox/synapseLoader.cpp

```diff
@@ -20,50 +20,72 @@
 #include "synapseLoader.h"
 #include "uriHandler.h"
 
-#include <brion/brion.h>
-#include <lunchbox/os.h>
-#include <lunchbox/memoryMap.h>
-#include <boost/progress.hpp>
-#ifdef final
-#  undef final
-#endif
+#include <brain/brain.h>
+
+#include <future>
+
 
 namespace fivox
 {
+
 class SynapseLoader::Impl
 {
 public:
     Impl( EventSource& output, const URIHandler& params )
         : _output( output )
+        , _circuit( params.getConfig( ))
+        , _preGIDs( params.getPreGIDs( ))
+        , _postGIDs( params.getGIDs( ))
+        , _synapses( _loadSynapseStream( ))
+        , _numChunks( _synapses.getRemaining( ))
+    {
+        _future = _synapses.read();
+
+        // compute circuit bounding box as we don't have any synapses at this
+        // point
+        const auto& gids = _circuit.getGIDs();
+        const brion::Vector3fs& positions = _circuit.getPositions( gids );
+
+        AABBf bbox;
+        for( const auto& position : positions )
+            bbox.merge( position );
+        _output.setBoundingBox( bbox );
+    }
+
+    brain::SynapsesStream _loadSynapseStream()
     {
-        const auto& gids = params.getGIDs();
-
-        LBINFO << "Loading synapses for " << gids.size() << " cells..."
-               << std::endl;
-        boost::progress_display progress( gids.size( ));
-        const brion::Synapse synapses( params.getConfig().getSynapseSource().getPath() +
-                                       "/nrn_positions.h5" );
-
-        std::vector< Vector3f > positions;
-        for( const uint32_t gid : gids )
-        {
-            const brion::SynapseMatrix& data =
-                synapses.read( gid, brion::SYNAPSE_PRESYNAPTIC_SURFACE_X |
-                                    brion::SYNAPSE_PRESYNAPTIC_SURFACE_Y |
-                                    brion::SYNAPSE_PRESYNAPTIC_SURFACE_Z );
-            for( size_t i = 0; i < data.shape()[0]; ++i )
-                positions.emplace_back(
-                        Vector3f( data[i][0], data[i][1], data[i][2] ));
-            ++progress;
-        }
-
-        _output.resize( positions.size( ));
-        for( size_t i = 0; i < positions.size(); ++i )
-            _output.update( i, positions.data()[i], /*radius*/0.f, /*val*/ 1.f);
+        if( _preGIDs.empty( ))
+            return _circuit.getAfferentSynapses( _postGIDs,
+                                            brain::SynapsePrefetch::positions );
+
+        return _circuit.getProjectedSynapses( _preGIDs, _postGIDs,
+                                            brain::SynapsePrefetch::positions );
+    }
+
+    ssize_t load( const size_t /*chunkIndex*/, const size_t numChunks )
+    {
```

Comment:
Hmm, I only see to have blocking load here then. It's still out-of-core, but no more prefetching. Maybe that's better and more correct for now. I'll test if it has any negative performance implications.